### PR TITLE
Pull request to correct dysfunctional tests

### PR DESCRIPTION
**Pull request by bholagabbar**
Dkayiwa: Hope you can figure out the problems locally!

### CODE REVIEW

File: omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java [hunk 1]
```diff
@@ -123,12 +125,20 @@ protected ModelAndView onSubmit(HttpServletRequest request, HttpServletResponse
...
 				}
-				
-				if (user == null /*|| StringUtils.isEmpty(user.getSecretQuestion())*/) {
+
+				String secretQuestion = null;
+				if(user != null) {
+					secretQuestion = Context.getUserService().getSecretQuestion(user);
+				}
+
+				if (user == null) {
```
Comment (wluyima) [inline omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java]: You should probably merge these if clauses

File: omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java [hunk 2]
```diff
@@ -147,41 +157,103 @@ protected ModelAndView onSubmit(HttpServletRequest request, HttpServletResponse
...
-				} else if (/*user.getSecretQuestion() != null &&*/ Context.getUserService().isSecretAnswer(user, secretAnswer)) {
+				} else if (secretQuestion != null && Context.getUserService().isSecretAnswer(user, secretAnswer)) {
 					
 					StringBuilder randomPassword = new StringBuilder();
-					for (int i = 0; i < 8; i++) {
-						randomPassword.append(String.valueOf((Math.random() * (127 - 48) + 48)));
+					ArrayList<Character> buildRandomPassword = new ArrayList<Character>();
+
+					//Ensure atleast 1 uppercase, 1 lowercase as well as 1 digit in password (Otherwise, exception is thrown)
+					buildRandomPassword.add((char) (65 + (int) Math.abs(Math.random()) % 26));
```
Comment (wluyima) [inline omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java]: You can use RandomStringUtils to generate an alphanumeric string and an upper case letter and concatenate them instead of doing this yourself

File: omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ * Test the different aspects of
+ * {@link org.openmrs.web.controller.ForgotPasswordFormController}
+ */
+public class ForgotPasswordFormControllerTest extends BaseModuleWebContextSensitiveTest {
+
+	protected static final String TEST_DATA= "org/openmrs/web/controller/include/ForgotPasswordFormControllerTest.xml";
+
+	@Before
+	public void updateSearchIndex() {
+		super.updateSearchIndex();
```
Comment (wluyima) [inline omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java]: Why do you need this? This is only required for concepts

File: omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	/**
+	 * Check to see if the admin's secret question comes back
+	 *
+	 * @throws Exception
+	 */
+	@Test
+	public void shouldNotNotFailOnNotFoundUsernameTEST() throws Exception {
+		executeDataSet(TEST_DATA);
```
Comment (wluyima) [inline omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java]: If your test still fails, it could be because you are adding this test data which i think duplicates the standard dataset

Comment (wluyima) [inline omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java]: I suggest you get rid of it and set the secret question for one of the users in the standard test dataset and use that before the test

Comment (bholagabbar) [inline omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java]: Ok. How do I include the standard dataset? Is it already inherited?

File: omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java (hunk 3 above)
Comment (bholagabbar) [inline omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java]: Ok I was just trying out new things..seeing if it works

File: omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java (hunk 2 above)
Comment (bholagabbar) [inline omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java]: Yes, but again, that would involve some conversion etc...I will do it if you say so

File: omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java (hunk 1 above)
Comment (bholagabbar) [inline omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java]: yes I will do that. Sorry you have had to repeat yourself. I was just focusing on tests

File: omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java (hunk 4 above)
Comment (wluyima) [inline omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java]: It's include by default, so you don't have to do anything

Comment (wluyima) [inline omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java]: As long as your test is a subclass of BaseContextSensitiveTest which BaseModuleWebContextSensitiveTest is and in turn is your test class is already

File: omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java (hunk 2 above)
Comment (wluyima) [inline omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java]: What conversion? All you have to is:

`RandomStringUtils.randomAlphabetic(1).toUpperCase() + RandomStringUtils.randomAlphanumeric(minChar)`

minChar should be the value of the GP_PASSWORD_MINIMUM_LENGTH global property value, default value should be 7(technically 8) but the eighth letter is the random uppercase letter we are prepending to it
